### PR TITLE
feat(ops): add Redis pressure panel with memory % and engine CPU

### PR DESCRIPTION
**Pull request by rogeriochaves**
## Summary

During the 2026-05-21 incident the Redis main thread sat at **100% for 5+ hours** while `/ops` exposed **no Redis pressure signal**:

- Memory was technically collected by the backend but only surfaced as a tiny `sublabel` under the "DLQ" tile.
- Engine CPU was never collected.
- Investigators had to use the Redis CLI on a replica to see what was happening.

This PR adds a first-class **"Redis pressure"** panel below the throughput strip, with three stats:

| Stat | Source | Warning color |
|---|---|---|
| Memory | `used_memory_human` / `maxmemory` + computed percent | ≥ 80% |
| Engine CPU | Diff of `used_cpu_user_main_thread` + `used_cpu_sys_main_thread` between two collect cycles | ≥ 70% |
| Connections | `connected_clients` | — |

The legacy `data.redisMemoryUsed` sublabel under "DLQ" is removed in favor of the real panel.

### Healthy — prod-like numbers (3.2G / 10.4G memory, 12.3% engine CPU)
healthy

### First cycle — CPU sample not ready yet
The engine-CPU percent is derived from two successive INFO snapshots, so on the very first cycle (and on the cycle after a Redis restart) we don't have enough data. Instead of guessing or showing `0%`, we render `-` with a `sampling…` sublabel.

sampling

### Saturated — 95% memory, 98% engine CPU (what May 21 would have looked like)
Memory and engine CPU both render in the warning color (`red.500`).

saturated

### `maxmemory = 0` (unlimited) — graceful fallback
Without a configured max, we omit the percent rather than showing `Infinity%`.

unlimited

> Screenshots captured by `src/components/ops/dashboard/__tests__/RedisPressurePanel.browser.test.tsx` rendering the real component inside Chromium via Playwright; reproducible with `pnpm test:browser`.

## Why engine CPU specifically

Redis processes commands on a single thread, so CloudWatch's `EngineCPUUtilization` is the most diagnostic metric during saturation incidents. We diff the **main-thread** cumulative counters between collect cycles — same math CloudWatch does — so this is the same number you'd see in AWS.

## The pure helper handles the tricky cases

`src/server/app-layer/ops/redis-engine-cpu.ts` exports `computeEngineCpuPercent({ prev, nextUserSec, nextSysSec, nextSampledAt })`. It returns `null` when:

- There is no previous sample yet (first collection cycle → UI shows "-" with "sampling…" sublabel)
- The cumulative counter went backwards (Redis restarted → would otherwise display a huge negative percent)
- Two samples were taken at the same instant (would divide by zero)

Otherwise it returns the percent rounded to one decimal.

## Specs and tests (no mocks where it matters)

Behavior spec: `specs/ops/redis-pressure.feature`

- **Unit (8 tests)** on `redis-engine-cpu.ts`: pure CPU-percent math — first-cycle null, rounding to 1dp, counter rewind, same-instant samples.
- **Integration (3 tests)** against a real Redis container via the project's existing testcontainers infra: parses `INFO`, takes two samples 50ms+ apart with 10K real ops in between, derives a non-negative percent ≤ 105% (allows for clock skew).
- **RTL (8 tests)** rendering the panel inside a real ChakraProvider: layout, percent rounding, the "sampling…" placeholder, the unlimited-maxmemory fallback, and threshold colors.
- **Browser (4 captures)** in real Chromium via Playwright: produces the 4 screenshots embedded above.

## Boy-scout cleanup

Pre-existing biome `noFloatingPromises` warning at `OpsMetricsCollector.start()` (the first `this.collect()` call was unawaited). Fixed with explicit `void` to preserve fire-and-forget semantics without blocking startup.

## Test plan

- [x] `pnpm typecheck` — clean
- [x] `pnpm exec biome check` on changed files — clean (1 pre-existing warning fixed)
- [x] `pnpm test:unit src/server/app-layer/ops` — 57/57 pass
- [x] `pnpm test:integration` for the two new integration files — 11/11 pass
- [x] `pnpm test:browser` for the new browser test — 4/4 pass (PNGs produced)
- [x] Screenshots pushed to `langwatch/pr-screenshots` and embedded above
- [ ] After merge: verify the live panel on lw-dev `/ops` matches the captured screenshots

### CODE REVIEW

**Comment by rogeriochaves** [comment]
@coderabbitai addressed all three findings in ca7945e:

- **Unused `VStack` import** (RedisPressurePanel.tsx): removed.
- **Spec inconsistency for unlimited maxmemory** (redis-pressure.feature line 41): scenario text now reads "memory percent is omitted instead of showing 'Infinity%'" to match the panel + integration test contract.
- **Warning-color tests not asserting style** (RedisPressurePanel.integration.test.tsx ~89-99, ~119-126): replaced the brittle `style.color` check (Chakra v3 paints via CSS vars on classes, so inline style stays empty) with a stable `data-warning="true|false"` attribute on the memory + CPU stat tiles. Added two paired below-threshold tests so a future regression that drops the warning condition would fail. RTL: 10/10 pass locally.

File: langwatch/specs/ops/redis-pressure.feature [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    Given the dashboard renders
+    When I inspect the DLQ tile
+    Then the DLQ tile no longer carries a Redis memory string as sublabel
+
+  @integration
+  Scenario: Redis stats appear inline with the throughput/latency tiles
+    Given the dashboard renders
+    When I view the top stat strip
+    Then Redis mem, Redis CPU, and Redis conns tiles appear alongside Staged/s, Completed/s, etc.
+    And the tiles wrap to a second row when the viewport is too narrow
```
[the remark this answers is not in the document]
Comment (rogeriochaves) [reply langwatch/specs/ops/redis-pressure.feature:109]: Intentional: `Engine CPU` is the metric concept (used in every scenario referencing the value), `Redis CPU` is the literal UI tile label in the inline strip, alongside `Redis mem` and `Redis conns`. Calling the inline tile `Engine CPU` would break parallelism with its neighbors. Leaving the layout scenario as `Redis CPU` because it specifically describes what the user sees on screen.

**Comment by rogeriochaves** [comment]
@coderabbitai round 2 addressed in 0f137b40a:

- **maxmemory mismatch** (spec): aligned setup to 9.69GB to match the rendered sublabel.
- **rounding-before-threshold** (RedisStatTiles.tsx): split raw vs rounded, threshold now uses raw, display still rounds to one decimal. Added test for the 79.95% edge case.
- **terminology** (spec layout scenario): kept as "Redis CPU" intentionally. Engine CPU = the metric concept (used in every scenario about the value), Redis CPU = literal UI tile label parallel to Redis mem / Redis conns.

File: langwatch/specs/ops/redis-pressure.feature [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    Given Redis reports used_memory=8.84GB and maxmemory=9.69GB
+    When the dashboard loads
+    Then the Redis mem tile is rendered in the warning color
+
+  @unit
+  Scenario: Memory warning uses the raw ratio so 79.95% does not round up to 80%
+    Given Redis reports used_memory and maxmemory at a ratio of 79.95%
+    When the dashboard data is built
+    Then the Redis mem tile is NOT rendered in the warning color
+    And the displayed percent is 80% (rounded to one decimal)
```
[the remark this answers is not in the document]
Comment (rogeriochaves) [reply langwatch/specs/ops/redis-pressure.feature:39]: Not changing this one: `Math.round(79.95 * 10) / 10` evaluates to the integer `80` (not `80.0`), and JS interpolates that as `"80%"`. Asserting `"80.0%"` would actually fail the test. The spec accurately describes what the UI renders.
